rclcpp/src/rclcpp/time.cpp

```diff
@@ -0,0 +1,233 @@
+  return time_point;
+}
+
+}  // namespace
+
+namespace rclcpp
+{
+
+Time
+Time::now(rcl_time_source_type_t clock)
+{
+  // TODO(karsten1987): This impl throws explicitely on RCL_ROS_TIME
+  // we have to do this, because rcl_time_source_init returns RCL_RET_OK
+  // for RCL_ROS_TIME, however defaults to system time under the hood.
+  // ref: https://github.com/ros2/rcl/blob/master/rcl/src/rcl/time.c#L66-L77
+  // This section can be removed when rcl supports ROS_TIME correctly.
+  if (clock == RCL_ROS_TIME) {
+    throw std::runtime_error("RCL_ROS_TIME is currently not implemented.");
+  }
+
+  Time now(0, 0, clock);
+
+  auto ret = rcl_time_point_get_now(&now.rcl_time_);
+  if (ret != RCL_RET_OK) {
+    rclcpp::exceptions::throw_from_rcl_error(
+      ret, "could not get current time stamp: ");
+  }
+
+  return now;
+}
+
+Time::Time(int32_t seconds, uint32_t nanoseconds, rcl_time_source_type_t clock)
+: rcl_time_source_(init_time_source(clock)),
+  rcl_time_(init_time_point(rcl_time_source_))
+{
+  if (seconds < 0) {
+    throw std::runtime_error("can't convert a negative time msg to rclcpp::Time");
+  }
+
+  uint64_t ns = RCL_S_TO_NS(static_cast<uint64_t>(seconds));
+  ns += nanoseconds;
+  rcl_time_.nanoseconds = ns;
+}
+
+Time::Time(uint64_t nanoseconds, rcl_time_source_type_t clock)
+: rcl_time_source_(init_time_source(clock)),
+  rcl_time_(init_time_point(rcl_time_source_))
+{
+  rcl_time_.nanoseconds = nanoseconds;
+}
+
+Time::Time(const builtin_interfaces::msg::Time & time_msg)  // NOLINT
+: rcl_time_source_(init_time_source(RCL_SYSTEM_TIME)),
+  rcl_time_(init_time_point(rcl_time_source_))
+{
+  if (time_msg.sec < 0) {
+    throw std::runtime_error("can't convert a negative time msg to rclcpp::Time");
+  }
+
+  rcl_time_.nanoseconds = RCL_S_TO_NS(static_cast<uint64_t>(time_msg.sec));
+  rcl_time_.nanoseconds += time_msg.nanosec;
+}
+
+Time::~Time()
+{
+  if (rcl_time_point_fini(&rcl_time_) != RCL_RET_OK) {
+    RCUTILS_LOG_FATAL("failed to reclaim rcl_time_point_t in destructor of rclcpp::Time")
+  }
+}
+
+Time::operator builtin_interfaces::msg::Time() const
+{
+  builtin_interfaces::msg::Time msg_time;
+  msg_time.sec = static_cast<std::int32_t>(RCL_NS_TO_S(rcl_time_.nanoseconds));
+  msg_time.nanosec = static_cast<std::uint32_t>(rcl_time_.nanoseconds % (1000 * 1000 * 1000));
+  return msg_time;
+}
+
+void
+Time::operator=(const builtin_interfaces::msg::Time & time_msg)
+{
+  if (time_msg.sec < 0) {
+    throw std::runtime_error("can't convert a negative time msg to rclcpp::Time");
+  }
+
+  this->rcl_time_source_ = init_time_source();
+  this->rcl_time_ = init_time_point(this->rcl_time_source_);
+
+  rcl_time_.nanoseconds = RCL_S_TO_NS(static_cast<uint64_t>(time_msg.sec));
+  rcl_time_.nanoseconds += time_msg.nanosec;
+}
+
+bool
+Time::operator==(const rclcpp::Time & rhs) const
+{
+  if (rcl_time_.time_source->type != rhs.rcl_time_.time_source->type) {
+    throw std::runtime_error("can't compare times with different time sources");
```

Comment:
I think we need to have a better overview of how this class should work. For example, I think that the fact that you cannot compare two Time objects with different sources is a runtime error is very inconvenient. Ideally it would be a compiler error when you try to compare a "system based" time point and a "steady based" time point. This pattern makes more sense when you're abstracting "system based" time point from "ROS based" time point because they have the same reference point. We might need to rethink the way it is setup in `rcl` as well to reflect this.